src/pmem.c

```diff
@@ -60,34 +59,43 @@ void pmemThresholdInit(void)
             serverAssert(NULL);
     }
 }
-//TODO: modify logic to check the trend of actual Ratio using pmem_checkpoint_value and dram_checkpoint_value
+
 void adjustPmemThresholdCycle(void) {
+    double step;
     if (server.memory_alloc_policy == MEM_POLICY_RATIO) {
         run_with_period(server.ratio_check_period) {
+            step = 0.05;
             size_t pmem_memory = zmalloc_used_pmem_memory();
             size_t dram_memory = zmalloc_used_memory();
-            size_t total_memory = pmem_memory + dram_memory;
-            size_t total_memory_checkpoint = server.pmem_checkpoint_value + server.dram_checkpoint_value;
+            size_t total_memory_actual = pmem_memory + dram_memory;
             // do not modify threshold when change in memory usage is too small
-            if (absDiff(total_memory_checkpoint, total_memory) > 100) {
+            if (absDiff(server.total_memory_checkpoint, total_memory_actual) > 100) {
+
                 //revert logic to avoid division by zero
-                double setting_state = (double)server.dram_pmem_ratio.pmem_val/server.dram_pmem_ratio.dram_val;
-                double current_state = (double)pmem_memory/dram_memory;
-                size_t threshold = zmalloc_get_threshold();
-                if (fabs(setting_state-current_state) > 0.1) {
-                    if (setting_state < current_state) {
-                        size_t higher_threshold = THRESHOLD_UP(threshold);
-                        if (higher_threshold > server.dynamic_threshold_max) return;
-                        zmalloc_set_threshold(higher_threshold);
+                double target_ratio = (double)server.dram_pmem_ratio.pmem_val/server.dram_pmem_ratio.dram_val;
+                double current_ratio = (double)pmem_memory/dram_memory;
+                double current_ratio_diff = fabs(current_ratio - target_ratio);
+                if (fabs(target_ratio-current_ratio) > 0.02) {
+                    //current ratio is worse than moment before
+                    if (current_ratio_diff > server.ratio_diff_checkpoint)
+                    {
+                        //increase threshold step
+                        step *= 5;
+                    }
+                    size_t threshold = zmalloc_get_threshold();
```

Comment:
The following logic should be under the if (current_ratio_diff > server.ratio_diff_checkpoint) { }, since if the current ratio is better than moment before, we don't need to do any adjustmnent on the threshold.

Reply:
_"if the current ratio is better than moment before, we don't need to do any adjustmnent on the threshold."_
In such a case, we can have a situation the current ratio is better than previous, but it is still far from target one and the trend of change is very slow. By continuously increasing/decreasing threshold we accelerate reaching target ratio. 

_Suggest: if target_ratio-current_ratio bigger, we can adjust the step bigger, if smaller, we can adjust the step smaller. This can help to ratio to convergency quicker._
I can give a try, however, I saw that the biggest impact on threshold and ratio comes from checking the trend of ratio change:
Now there are two rates of threshold modification:
- faster - when the current ratio is worse than the previous
- slower - when the current ratio is better than the previous.